File: Task/XXXX-redacted/Julia/xxxx-redacted.jl
function doif_equals(word, pattern, insens=false)
    regex = insens ? Regex("^$pattern\$", "i") : Regex("^$pattern\$")
    return replace(word, regex => pattern in multichars ? "X" : "X"^length(pattern))
end
doif_ci_equals(word, pattern) = doif_equals(word, pattern, true)

function doif_includes(word, pattern, insens=false)
    regex = insens ? Regex(pattern, "i") : Regex(pattern)
    return replace(word, regex => "X"^length(pattern))
end
doif_ci_includes(word, pattern) = doif_includes(word, pattern, true)

function overkill(word, pattern, insens=false)
    regex = insens ? Regex(pattern, "i") : Regex(pattern)
    return occursin(regex, word) ? "X"^length(word) : word
end
ci_overkill(word, pattern) = overkill(word, pattern, true)

const method = Dict(
    "[w|s|n]" => doif_equals,
    "[w|i|n]" => doif_ci_equals,
    "[p|s|n]" => doif_includes,
    "[p|i|n]" => doif_ci_includes,
    "[p|s|o]" => overkill,
    "[p|i|o]" => ci_overkill
)
const multichars = Set(["👨‍👩‍👦", ])

function redact(teststring, pattern)
    ws = split(teststring, r"[^ \?\"\.]+")
    words = filter(!=(""), split(teststring, r"[\s\?\"\.]+"))
    fs = popfirst!(words)
    f = method[fs]
    return fs * ws[2] * mapreduce(i -> f(words[i], pattern) * ws[i + 2], *, 1:length(words))
end

const testtext = """
[w|s|n] Tom? Toms bottom tomato is in his stomach while playing the "Tom-tom" brand tom-toms. That's so tom.
[w|i|n] Tom? Toms bottom tomato is in his stomach while playing the "Tom-tom" brand tom-toms. That's so tom.
[p|s|n] Tom? Toms bottom tomato is in his stomach while playing the "Tom-tom" brand tom-toms. That's so tom.
[p|i|n] Tom? Toms bottom tomato is in his stomach while playing the "Tom-tom" brand tom-toms. That's so tom.
[p|s|o] Tom? Toms bottom tomato is in his stomach while playing the "Tom-tom" brand tom-toms. That's so tom.
[p|i|o] Tom? Toms bottom tomato is in his stomach while playing the "Tom-tom" brand tom-toms. That's so tom.
"""
const stretchtext = "[w|s|n] 🧑 👨 🧔 👨‍👩‍👦"

for test in [(testtext, ["Tom", "tom"]), (stretchtext, ["👨", "👨‍👩‍👦"])]
    for pat in test[2]
        println("\nRedact pattern \"$pat\":")
        for teststring in string.(split(strip(test[1]), r"\n"))
            println(redact(teststring, pat))
        end
    end
    println()
end

File: Task/XXXX-redacted/Python/xxxx-redacted.py
#!/usr/bin/env python3

def redact(source, word, partial=False, insensitive=False, overkill=False):
    def different(s, w):
        if insensitive:
            return s.upper() != w.upper()
        else:
            return s != w

    def is_word_char(c):
        return c == '-' or c.isalpha()

    temp = list(source)
    i = 0
    while i <= len(temp) - len(word):
        match = True
        j = 0
        while j < len(word):
            if different(temp[i + j], word[j]):
                match = False
                break
            j += 1
        if match:
            beg = i
            end = i + len(word)
            if not partial:
                if beg > 0 and is_word_char(temp[beg - 1]):
                    i += 1
                    continue
                if end < len(temp) and is_word_char(temp[end]):
                    i += 1
                    continue
            if overkill:
                while beg > 0 and is_word_char(temp[beg - 1]):
                    beg -= 1
                while end < len(temp) - 1 and is_word_char(temp[end]):
                    end += 1
            for k in range(beg, end):
                temp[k] = 'X'
        i += 1
    return ''.join(temp)

def example(source, word):
    print("Redact '", word, "':")
    print("[w|s|n]", redact(source, word, False, False, False))
    print("[w|i|n]", redact(source, word, False, True, False))
    print("[p|s|n]", redact(source, word, True, False, False))
    print("[p|i|n]", redact(source, word, True, True, False))
    print("[p|s|o]", redact(source, word, True, False, True))
    print("[p|i|o]", redact(source, word, True, True, True))
    print()

if __name__ == "__main__":
    text = 'Tom? Toms bottom tomato is in his stomach while playing the "Tom-tom" brand tom-toms. That\'s so tom'
    example(text, "Tom")
    example(text, "tom")
